client/parser/format.go

```diff
@@ -0,0 +1,748 @@
+					case SharpNode:
+						pitchText.WriteString("+")
+					}
+				}
+			}
+
+			slurText := ""
+			for _, child := range node.Children[1:] {
+				if child.Type == TieNode {
+					slurText = "~"
+				}
+			}
+
+			if len(node.Children) > 1 && node.Children[1].Type == DurationNode {
+				err = f.formatWithDuration(
+					pitchText.String(), node.Children[1], slurText,
+				)
+				if err != nil {
+					return err
+				}
+			} else {
+				f.write(fmt.Sprintf("%s%s", pitchText.String(), slurText))
+			}
+
+		case OctaveDownNode:
+			f.write("<")
+
+		case OctaveSetNode:
+			f.write(fmt.Sprintf("o%d", node.Literal.(int32)))
+
+		case OctaveUpNode:
+			f.write(">")
+
+		case RepeatNode:
+			if err := node.expectNChildren(2); err != nil {
+				return err
+			}
+
+			err := f.formatInnerEvents(node.Children[0])
+			if err != nil {
+				return err
+			}
+
+			times, err := node.Children[1].expectNodeType(TimesNode)
+			if err != nil {
+				return err
+			}
+
+			f.write(fmt.Sprintf("*%d", times.Literal.(int32)))
+
+		case OnRepetitionsNode:
+			if err := node.expectNChildren(2); err != nil {
+				return err
+			}
+
+			err := f.formatInnerEvents(node.Children[0])
+			if err != nil {
+				return err
+			}
+
+			repetitions, err := node.Children[1].expectNodeType(RepetitionsNode)
+			if err != nil {
+				return err
+			}
+
+			ranges := []string{}
+			for _, child := range repetitions.Children {
+				rr, err := child.expectNodeType(RepetitionRangeNode)
+				if err != nil {
+					return err
+				}
+
+				if err := rr.expectNChildren(2); err != nil {
+					return err
+				}
+
+				fr, err := rr.Children[0].expectNodeType(FirstRepetitionNode)
+				if err != nil {
+					return err
+				}
+
+				lr, err := rr.Children[1].expectNodeType(LastRepetitionNode)
+				if err != nil {
+					return err
+				}
+
+				frNum := fr.Literal.(int32)
+				lrNum := lr.Literal.(int32)
+
+				if frNum == lrNum {
+					ranges = append(ranges,
+						fmt.Sprintf("%d", frNum),
+					)
+				} else {
+					ranges = append(ranges,
+						fmt.Sprintf("%d-%d", frNum, lrNum),
+					)
+				}
+			}
+			f.write(fmt.Sprintf("'%s", strings.Join(ranges, ",")))
+
+		case RestNode:
+			if len(node.Children) > 0 {
+				duration, err := node.Children[0].expectNodeType(DurationNode)
+				if err != nil {
+					return err
+				}
+
+				err = f.formatWithDuration("r", duration, "")
+				if err != nil {
+					return err
+				}
+			} else {
+				f.write("r")
+			}
+
+		case VariableDefinitionNode:
+			// Variable definitions are incredibly tricky to formatInnerEvents because
+			// formatted text must be on the same line as the variable name.
+			// We handle this by maintaining varDefState:
+			// - While "None", behaviour is normal.
+			// - While "Defining", we never flush/wrap to a new line.
+			// - While "LastNode", we allow event sequences (including repeats)
+			// 	 and voice groups to indent and continue on new lines.
+			// 	 This is complicated, but any alternative I tried was worse.
```

Comment:
I can't think of any alternatives at the moment, but I think this is perfectly fine for now.